Empty target groups from category dtypes

When the target column is a category dtype, the label search can come across empty target groups which results in an invalid offset error. A temporary workaround is using a different dtype for the target column. The long-term fix can be updating the label search to skip target groups with empty dataframes.

#### Example

```python
import composeml as cp
import pandas as pd

data = [
    {"time": "2014-01-01 03:05:15", "user_id": "4", "amount": "29.37"},
    {"time": "2014-01-01 07:51:15", "user_id": "2", "amount": "18.81"},
    {"time": "2014-01-01 05:01:10", "user_id": "5", "amount": "83.33"},
    {"time": "2014-01-01 00:45:30", "user_id": "1", "amount": "108.11"},
    {"time": "2014-01-01 00:43:20", "user_id": "4", "amount": "147.73"},
]

dtypes = dict(time="datetime64[ns]", user_id="category")
df = pd.DataFrame(data=data).astype(dtypes)
```

```
                 time user_id  amount
0 2014-01-01 03:05:15       4   29.37
1 2014-01-01 07:51:15       2   18.81
2 2014-01-01 05:01:10       5   83.33
3 2014-01-01 00:45:30       1  108.11
4 2014-01-01 00:43:20       4  147.73
```

```python
lm = cp.LabelMaker(
    target_entity="user_id",
    time_index="time",
    window_size=3,
    labeling_function=sum,
)

lt = lm.search(df.head(3).sort_values("time"), -1)
```

```
AssertionError: offset must be position or time based
```

--- a/composeml/label_maker.py
+++ b/composeml/label_maker.py
@@ -1,6 +1,7 @@
 from sys import stdout
 
 from pandas import Series
+from pandas.api.types import is_categorical_dtype
 from tqdm import tqdm
 
 from composeml.data_slice import DataSliceGenerator
@@ -205,6 +206,11 @@ class LabelMaker:
 
         df = self.set_index(df)
         total = search.expected_count if search.is_finite else 1
+        # If the target is categorical, make sure there are no unused categories
+        if is_categorical_dtype(df[self.target_dataframe_name]):
+            df[self.target_dataframe_name] = df[
+                self.target_dataframe_name
+            ].cat.remove_unused_categories()
         target_groups = df.groupby(self.target_dataframe_name)
         total *= target_groups.ngroups
 

--- a/composeml/tests/test_label_maker.py
+++ b/composeml/tests/test_label_maker.py
@@ -599,3 +599,16 @@ def test_minimum_data_per_group_error(transactions):
 
     with pytest.raises(ValueError, match=match):
         lm.search(transactions, 1, minimum_data=minimum_data)
+
+
+def test_label_maker_categorical_target_with_missing_data(transactions, total_spent_fn):
+    transactions = transactions.copy()
+    transactions["customer_id"] = transactions["customer_id"].astype("category")
+    lm = LabelMaker(
+        target_dataframe_name="customer_id",
+        time_index="time",
+        window_size=3,
+        labeling_function=total_spent_fn,
+    )
+    # use on only the first 8 rows so the df will not contain data for customer 3
+    lm.search(transactions.head(8), -1)

--- a/docs/source/release_notes.rst
+++ b/docs/source/release_notes.rst
@@ -5,13 +5,14 @@ Future Release
 ==============
     * Enhancements
     * Fixes
+        * Update to avoid error with a categorical target with unused categories (:pr:`349`)
     * Changes
     * Documentation Changes
     * Testing Changes
         * Add create feedstock PR workflow (:pr:`346`)
 
     Thanks to the following people for contributing to this release:
-    :user:`gsheni`
+    :user:`gsheni`, :user:`thehomebrewnerd`
     
 v0.9.1 Nov 2, 2022
 ==================

